// @gorniv/ngx-universal/esm2015/cookie/cookie-backend.service.js
import { __decorate, __metadata, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { CookieService } from './cookie.service';
import { CookieOptionsProvider } from './cookie-options-provider';
import { NgxRequest, NgxResponse } from './tokens';
import * as ɵngcc0 from '@angular/core';
let CookieBackendService = class CookieBackendService extends CookieService {
    constructor(request, response, _optionsProvider) {
        super(_optionsProvider);
        this.request = request;
        this.response = response;
    }
    get cookieString() {
        return this.request.cookie || this.request.headers['cookie'] || '';
    }
    set cookieString(val) {
        this.request.cookie = val;
        this.response.cookie = val;
    }
    put(key, value, options = {}) {
        let findKey = false;
        let newCookie = Object.keys(this.getAll())
            // tslint:disable-next-line: no-shadowed-variable
            .map((keyItem) => {
            if (keyItem === key) {
                findKey = true;
                return `${key}=${value}`;
            }
            return `${keyItem}=${this.get(keyItem)}`;
        })
            .join('; ');
        if (!findKey) {
            newCookie += `; ${key}=${value}`;
        }
        this.request.headers.cookie = newCookie;
        // not sure
        this.cookieString = newCookie;
    }
    remove(key, options) {
        const newCookie = Object.keys(this.getAll())
            // tslint:disable-next-line: no-shadowed-variable
            .map((keyItem) => {
            if (keyItem === key) {
                return '';
            }
            return `${keyItem}=${this.get(keyItem)}`;
        })
            .join('; ');
        this.request.headers.cookie = newCookie;
        // not sure
        this.cookieString = newCookie;
    }
};
CookieBackendService.ɵfac = function CookieBackendService_Factory(t) { return new (t || CookieBackendService)(ɵngcc0.ɵɵinject(NgxRequest), ɵngcc0.ɵɵinject(NgxResponse), ɵngcc0.ɵɵinject(CookieOptionsProvider)); };
CookieBackendService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: CookieBackendService, factory: CookieBackendService.ɵfac });
CookieBackendService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [NgxRequest,] }] },
    { type: undefined, decorators: [{ type: Inject, args: [NgxResponse,] }] },
    { type: CookieOptionsProvider }
];
CookieBackendService = __decorate([ __param(0, Inject(NgxRequest)),
    __param(1, Inject(NgxResponse)),
    __metadata("design:paramtypes", [Object, Object, CookieOptionsProvider])
], CookieBackendService);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CookieBackendService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [NgxRequest]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [NgxResponse]
            }] }, { type: CookieOptionsProvider }]; }, null); })();
export { CookieBackendService };
