// @ngx-meta/core/esm5/meta.module.js
import * as tslib_1 from "tslib";
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { MetaGuard } from './meta.guard';
import { MetaLoader, MetaStaticLoader } from './meta.loader';
import { MetaService } from './meta.service';
import * as ɵngcc0 from '@angular/core';
export var metaFactory = function () { return new MetaStaticLoader(); };
var MetaModule = (function () {
    function MetaModule(parentModule) {
        if (parentModule) {
            throw new Error('MetaModule already loaded; import in root module only.');
        }
    }
    MetaModule_1 = MetaModule;
    MetaModule.forRoot = function (configuredProvider) {
        if (configuredProvider === void 0) { configuredProvider = {
            provide: MetaLoader,
            useFactory: metaFactory
        }; }
        return {
            ngModule: MetaModule_1,
            providers: [configuredProvider, MetaGuard, MetaService]
        };
    };
    var MetaModule_1;
    MetaModule.ctorParameters = function () { return [
        { type: MetaModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
    ]; };
    MetaModule = MetaModule_1 = tslib_1.__decorate([ tslib_1.__param(0, Optional()), tslib_1.__param(0, SkipSelf()),
        tslib_1.__metadata("design:paramtypes", [MetaModule])
    ], MetaModule);
MetaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MetaModule });
MetaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MetaModule_Factory(t) { return new (t || MetaModule)(ɵngcc0.ɵɵinject(MetaModule, 12)); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MetaModule, [{
        type: NgModule
    }], function () { return [{ type: MetaModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();
    return MetaModule;
}());
export { MetaModule };
