// @ngx-meta/core/esm5/meta.guard.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { MetaService } from './meta.service';
import * as ɵngcc0 from '@angular/core';
var MetaGuard = (function () {
    function MetaGuard(meta) {
        this.meta = meta;
    }
    MetaGuard.prototype.canActivate = function (route, state) {
        var url = state.url;
        var metaSettings = route.hasOwnProperty('data') ? route.data.meta : undefined;
        this.meta.update(url, metaSettings);
        return true;
    };
    MetaGuard.prototype.canActivateChild = function (route, state) {
        return this.canActivate(route, state);
    };
    MetaGuard.ctorParameters = function () { return [
        { type: MetaService }
    ]; };
    MetaGuard = tslib_1.__decorate([ tslib_1.__metadata("design:paramtypes", [MetaService])
    ], MetaGuard);
MetaGuard.ɵfac = function MetaGuard_Factory(t) { return new (t || MetaGuard)(ɵngcc0.ɵɵinject(MetaService)); };
MetaGuard.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: MetaGuard, factory: function (t) { return MetaGuard.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MetaGuard, [{
        type: Injectable
    }], function () { return [{ type: MetaService }]; }, null); })();
    return MetaGuard;
}());
export { MetaGuard };
