// @ngx-meta/core/esm2015/meta.module.js
import * as ɵngcc0 from '@angular/core';
var MetaModule_1;
import * as tslib_1 from "tslib";
import { NgModule, Optional, SkipSelf } from '@angular/core';
import { MetaGuard } from './meta.guard';
import { MetaLoader, MetaStaticLoader } from './meta.loader';
import { MetaService } from './meta.service';
export const metaFactory = () => new MetaStaticLoader();
let MetaModule = MetaModule_1 = class MetaModule {
    constructor(parentModule) {
        if (parentModule) {
            throw new Error('MetaModule already loaded; import in root module only.');
        }
    }
    static forRoot(configuredProvider = {
        provide: MetaLoader,
        useFactory: metaFactory
    }) {
        return {
            ngModule: MetaModule_1,
            providers: [configuredProvider, MetaGuard, MetaService]
        };
    }
};
MetaModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: MetaModule });
MetaModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function MetaModule_Factory(t) { return new (t || MetaModule)(ɵngcc0.ɵɵinject(MetaModule, 12)); } });
MetaModule.ctorParameters = () => [
    { type: MetaModule, decorators: [{ type: Optional }, { type: SkipSelf }] }
];
MetaModule = MetaModule_1 = tslib_1.__decorate([ tslib_1.__param(0, Optional()), tslib_1.__param(0, SkipSelf()),
    tslib_1.__metadata("design:paramtypes", [MetaModule])
], MetaModule);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MetaModule, [{
        type: NgModule
    }], function () { return [{ type: MetaModule, decorators: [{
                type: Optional
            }, {
                type: SkipSelf
            }] }]; }, null); })();
export { MetaModule };
