// @ngx-meta/core/esm2015/meta.guard.js
import * as tslib_1 from "tslib";
import { Injectable } from '@angular/core';
import { MetaService } from './meta.service';
import * as ɵngcc0 from '@angular/core';
let MetaGuard = class MetaGuard {
    constructor(meta) {
        this.meta = meta;
    }
    canActivate(route, state) {
        const url = state.url;
        const metaSettings = route.hasOwnProperty('data') ? route.data.meta : undefined;
        this.meta.update(url, metaSettings);
        return true;
    }
    canActivateChild(route, state) {
        return this.canActivate(route, state);
    }
};
MetaGuard.ɵfac = function MetaGuard_Factory(t) { return new (t || MetaGuard)(ɵngcc0.ɵɵinject(MetaService)); };
MetaGuard.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: MetaGuard, factory: MetaGuard.ɵfac });
MetaGuard.ctorParameters = () => [
    { type: MetaService }
];
MetaGuard = tslib_1.__decorate([ tslib_1.__metadata("design:paramtypes", [MetaService])
], MetaGuard);
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(MetaGuard, [{
        type: Injectable
    }], function () { return [{ type: MetaService }]; }, null); })();
export { MetaGuard };
