// @nguniversal/express-engine/esm2015/src/main.js
/**
 * @fileoverview added by tsickle
 * Generated from: modules/express-engine/src/main.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as fs from 'fs';
import { ɵCommonEngine as CommonEngine } from '@nguniversal/common/engine';
import { REQUEST, RESPONSE } from '@nguniversal/express-engine/tokens';
/**
 * These are the allowed options for the engine
 * @record
 */
export function NgSetupOptions() { }
if (false) {
    /** @type {?} */
    NgSetupOptions.prototype.bootstrap;
    /** @type {?|undefined} */
    NgSetupOptions.prototype.providers;
}
/**
 * These are the allowed options for the render
 * @record
 */
export function RenderOptions() { }
if (false) {
    /** @type {?} */
    RenderOptions.prototype.req;
    /** @type {?|undefined} */
    RenderOptions.prototype.res;
    /** @type {?|undefined} */
    RenderOptions.prototype.url;
    /** @type {?|undefined} */
    RenderOptions.prototype.document;
}
/**
 * This holds a cached version of each index used.
 * @type {?}
 */
const templateCache = {};
/**
 * This is an express engine for handling Angular Applications
 * @param {?} setupOptions
 * @return {?}
 */
export function ngExpressEngine(setupOptions) {
    /** @type {?} */
    const engine = new CommonEngine(setupOptions.bootstrap, setupOptions.providers);
    return (/**
     * @param {?} filePath
     * @param {?} options
     * @param {?} callback
     * @return {?}
     */
    function (filePath, options, callback) {
        try {
            if (!setupOptions.bootstrap && !options.bootstrap) {
                throw new Error('You must pass in a NgModule or NgModuleFactory to be bootstrapped');
            }
            /** @type {?} */
            const req = options.req;
            /** @type {?} */
            const renderOptions = Object.assign({}, options);
            renderOptions.url =
                options.url || `${req.protocol}://${(req.get('host') || '')}${req.originalUrl}`;
            renderOptions.document = options.document || getDocument(filePath);
            renderOptions.providers = options.providers || [];
            renderOptions.providers =
                renderOptions.providers.concat(getReqResProviders(options.req, options.res));
            engine.render(renderOptions)
                .then((/**
             * @param {?} html
             * @return {?}
             */
            html => callback(null, html)))
                .catch(callback);
        }
        catch (err) {
            callback(err);
        }
    });
}
/**
 * Get providers of the request and response
 * @param {?} req
 * @param {?=} res
 * @return {?}
 */
function getReqResProviders(req, res) {
    /** @type {?} */
    const providers = [
        {
            provide: REQUEST,
            useValue: req
        }
    ];
    if (res) {
        providers.push({
            provide: RESPONSE,
            useValue: res
        });
    }
    return providers;
}
/**
 * Get the document at the file path
 * @param {?} filePath
 * @return {?}
 */
function getDocument(filePath) {
    return templateCache[filePath] = templateCache[filePath] || fs.readFileSync(filePath).toString();
}
