// @gorniv/ngx-universal/esm5/cookie/cookie-backend.service.js
import { __decorate, __extends, __metadata, __param } from "tslib";
import { Inject, Injectable } from '@angular/core';
import { CookieService } from './cookie.service';
import { CookieOptionsProvider } from './cookie-options-provider';
import { NgxRequest, NgxResponse } from './tokens';
import * as ɵngcc0 from '@angular/core';
var CookieBackendService = /** @class */ (function (_super) {
    __extends(CookieBackendService, _super);
    function CookieBackendService(request, response, _optionsProvider) {
        var _this = _super.call(this, _optionsProvider) || this;
        _this.request = request;
        _this.response = response;
        return _this;
    }
    Object.defineProperty(CookieBackendService.prototype, "cookieString", {
        get: function () {
            return this.request.cookie || this.request.headers['cookie'] || '';
        },
        set: function (val) {
            this.request.cookie = val;
            this.response.cookie = val;
        },
        enumerable: true,
        configurable: true
    });
    CookieBackendService.prototype.put = function (key, value, options) {
        var _this = this;
        if (options === void 0) { options = {}; }
        var findKey = false;
        var newCookie = Object.keys(this.getAll())
            // tslint:disable-next-line: no-shadowed-variable
            .map(function (keyItem) {
            if (keyItem === key) {
                findKey = true;
                return key + "=" + value;
            }
            return keyItem + "=" + _this.get(keyItem);
        })
            .join('; ');
        if (!findKey) {
            newCookie += "; " + key + "=" + value;
        }
        this.request.headers.cookie = newCookie;
        // not sure
        this.cookieString = newCookie;
    };
    CookieBackendService.prototype.remove = function (key, options) {
        var _this = this;
        var newCookie = Object.keys(this.getAll())
            // tslint:disable-next-line: no-shadowed-variable
            .map(function (keyItem) {
            if (keyItem === key) {
                return '';
            }
            return keyItem + "=" + _this.get(keyItem);
        })
            .join('; ');
        this.request.headers.cookie = newCookie;
        // not sure
        this.cookieString = newCookie;
    };
    CookieBackendService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [NgxRequest,] }] },
        { type: undefined, decorators: [{ type: Inject, args: [NgxResponse,] }] },
        { type: CookieOptionsProvider }
    ]; };
    CookieBackendService = __decorate([ __param(0, Inject(NgxRequest)),
        __param(1, Inject(NgxResponse)),
        __metadata("design:paramtypes", [Object, Object, CookieOptionsProvider])
    ], CookieBackendService);
CookieBackendService.ɵfac = function CookieBackendService_Factory(t) { return new (t || CookieBackendService)(ɵngcc0.ɵɵinject(NgxRequest), ɵngcc0.ɵɵinject(NgxResponse), ɵngcc0.ɵɵinject(CookieOptionsProvider)); };
CookieBackendService.ɵprov = ɵngcc0.ɵɵdefineInjectable({ token: CookieBackendService, factory: function (t) { return CookieBackendService.ɵfac(t); } });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(CookieBackendService, [{
        type: Injectable
    }], function () { return [{ type: undefined, decorators: [{
                type: Inject,
                args: [NgxRequest]
            }] }, { type: undefined, decorators: [{
                type: Inject,
                args: [NgxResponse]
            }] }, { type: CookieOptionsProvider }]; }, null); })();
    return CookieBackendService;
}(CookieService));
export { CookieBackendService };
