// @nguniversal/common/esm2015/src/state-transfer-initializer/module.js
/**
 * @fileoverview added by tsickle
 * Generated from: modules/common/src/state-transfer-initializer/module.ts
 * @suppress {checkTypes,constantProperty,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { DOCUMENT } from '@angular/common';
import { APP_INITIALIZER, NgModule } from '@angular/core';
/**
 * @param {?} doc
 * @return {?}
 */
import * as ɵngcc0 from '@angular/core';
export function domContentLoadedFactory(doc) {
    return (/**
     * @return {?}
     */
    () => new Promise((/**
     * @param {?} resolve
     * @param {?} _reject
     * @return {?}
     */
    (resolve, _reject) => {
        /** @type {?} */
        const contentLoaded = (/**
         * @return {?}
         */
        () => {
            doc.removeEventListener('DOMContentLoaded', contentLoaded);
            resolve();
        });
        if (doc.readyState === 'complete' || doc.readyState === 'interactive') {
            resolve();
        }
        else {
            doc.addEventListener('DOMContentLoaded', contentLoaded);
        }
    })));
}
export class StateTransferInitializerModule {
}
StateTransferInitializerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: StateTransferInitializerModule });
StateTransferInitializerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function StateTransferInitializerModule_Factory(t) { return new (t || StateTransferInitializerModule)(); }, providers: [
        { provide: APP_INITIALIZER, multi: true, useFactory: domContentLoadedFactory, deps: [DOCUMENT] },
    ] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(StateTransferInitializerModule, [{
        type: NgModule,
        args: [{
                providers: [
                    { provide: APP_INITIALIZER, multi: true, useFactory: domContentLoadedFactory, deps: [DOCUMENT] },
                ]
            }]
    }], null, null); })();
