// @gorniv/ngx-universal/esm2015/link/link.service.js
// https://github.com/angular/angular/issues/15776#issuecomment-291862953
import { __decorate, __metadata, __param } from "tslib";
import { Injectable, RendererFactory2, ViewEncapsulation, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/common';
let LinkService = class LinkService {
    constructor(rendererFactory, document) {
        this.rendererFactory = rendererFactory;
        this.document = document;
    }
    /**
     * Inject the State into the bottom of the <head>
     */
    addTag(tag, forceCreation) {
        try {
            const renderer = this.rendererFactory.createRenderer(this.document, {
                id: '-1',
                encapsulation: ViewEncapsulation.None,
                styles: [],
                data: {},
            });
            const link = renderer.createElement('link');
            const head = this.document.head;
            const selector = this._parseSelector(tag);
            if (head === null) {
                throw new Error('<head> not found within DOCUMENT.');
            }
            Object.keys(tag).forEach((prop) => {
                return renderer.setAttribute(link, prop, tag[prop]);
            });
            // [TODO]: get them to update the existing one (if it exists) ?
            renderer.appendChild(head, link);
        }
        catch (e) {
            console.error('Error within linkService : ', e);
        }
    }
    _parseSelector(tag) {
        // Possibly re-work this
        const attr = tag.rel ? 'rel' : 'hreflang';
        return `${attr}="${tag[attr]}"`;
    }
};
LinkService.ctorParameters = () => [
    { type: RendererFactory2 },
    { type: undefined, decorators: [{ type: Inject, args: [DOCUMENT,] }] }
];
LinkService = __decorate([
    Injectable({
        providedIn: 'root',
    })
    /*
     * -- LinkService --        [Temporary]
     * @MarkPieszak
     *
     * Similar to Meta service but made to handle <link> creation for SEO purposes
     * -- NOTE: Soon there will be an overall DocumentService within Angular that handles Meta/Link everything
     */
    ,
    Injectable(),
    __param(1, Inject(DOCUMENT)),
    __metadata("design:paramtypes", [RendererFactory2, Object])
], LinkService);
export { LinkService };
