// @nguniversal/common/esm5/src/state-transfer-initializer/module.js
/**
 * @license
 * Copyright Google LLC All Rights Reserved.
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://angular.io/license
 */
import { __decorate } from "tslib";
import { DOCUMENT } from '@angular/common';
import { APP_INITIALIZER, NgModule } from '@angular/core';
import * as ɵngcc0 from '@angular/core';
export function domContentLoadedFactory(doc) {
    return function () { return new Promise(function (resolve, _reject) {
        var contentLoaded = function () {
            doc.removeEventListener('DOMContentLoaded', contentLoaded);
            resolve();
        };
        if (doc.readyState === 'complete' || doc.readyState === 'interactive') {
            resolve();
        }
        else {
            doc.addEventListener('DOMContentLoaded', contentLoaded);
        }
    }); };
}
var StateTransferInitializerModule = /** @class */ (function () {
    function StateTransferInitializerModule() {
    }
StateTransferInitializerModule.ɵmod = ɵngcc0.ɵɵdefineNgModule({ type: StateTransferInitializerModule });
StateTransferInitializerModule.ɵinj = ɵngcc0.ɵɵdefineInjector({ factory: function StateTransferInitializerModule_Factory(t) { return new (t || StateTransferInitializerModule)(); }, providers: [
        { provide: APP_INITIALIZER, multi: true, useFactory: domContentLoadedFactory, deps: [DOCUMENT] },
    ] });
/*@__PURE__*/ (function () { ɵngcc0.ɵsetClassMetadata(StateTransferInitializerModule, [{
        type: NgModule,
        args: [{
                providers: [
                    { provide: APP_INITIALIZER, multi: true, useFactory: domContentLoadedFactory, deps: [DOCUMENT] },
                ]
            }]
    }], function () { return []; }, null); })();
    return StateTransferInitializerModule;
}());
export { StateTransferInitializerModule };
